File: apps/android/src/main/kotlin/space/bitos/app/ui/create/GifPickerSheet.kt
package space.bitos.app.ui.create

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.GridItemSpan
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.painter.ColorPainter
import androidx.compose.ui.graphics.vector.rememberVectorPainter
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.platform.LocalContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import space.bitos.app.data.media.GifPickerStore
import coil.compose.AsyncImage
import space.bitos.app.ui.theme.AppIcons
import space.bitos.app.ui.theme.BitOSColors
import space.bitos.app.ui.theme.BitOSSpacing
import space.bitos.core.publish.CachedGifs
import space.bitos.core.publish.GifChoice
import space.bitos.core.publish.GifPickerContract

/**
 * APP-008 GIF picker (legacy Flutter `GifPickerSheet` / web
 * `GifPicker.svelte` parity): trending on open, 350 ms debounced search,
 * Recent tab (≤12, persisted), 24 h trending cache, "Load more"
 * pagination, three-column animated preview grid and the "Powered by Giphy"
 * footer.
 * Picking hands back the full-resolution URL — the composer embeds it as
 * a plain image URL. All rules run in shared `GifPickerContract`; this
 * sheet owns HTTP, tiles and persistence only.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun GifPickerSheet(
    onPick: (GifChoice) -> Unit,
    onDismiss: () -> Unit,
    /** Open on the STICKERS tab (transparent cut-outs — what meme layers
     *  want on top of the media; the composer default stays GIFs). */
    defaultStickers: Boolean = false,
) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val store = remember { GifPickerStore(context) }
    val client = remember {
        OkHttpClient.Builder().callTimeout(10, java.util.concurrent.TimeUnit.SECONDS).build()
    }

    var query by remember { mutableStateOf("") }
    var items by remember { mutableStateOf<List<GifChoice>>(emptyList()) }
    var trendingSnapshot by remember { mutableStateOf<List<GifChoice>>(emptyList()) }
    var trendingSavedAt by remember { mutableStateOf(0L) }
    // Web parity: GIFs vs Stickers — Giphy's sticker endpoints return
    // transparent cut-outs. Each kind keeps its own trending snapshot.
    var stickersTab by remember { mutableStateOf(defaultStickers) }
    var stickersSnapshot by remember { mutableStateOf<List<GifChoice>>(emptyList()) }
    var stickersSavedAt by remember { mutableStateOf(0L) }
    var recent by remember { mutableStateOf<List<GifChoice>>(emptyList()) }
    var recentTab by remember { mutableStateOf(false) }
    var loading by remember { mutableStateOf(false) }
    var loadingMore by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf("") }
    var nextOffset by remember { mutableStateOf(0) }
    var hasMore by remember { mutableStateOf(true) }
    var restored by remember { mutableStateOf(false) }
    // Debounce guard: only the newest request may apply its results.
    var generation by remember { mutableStateOf(0) }

    fun persist() {
        store.save(
            CachedGifs(
                recent = recent,
                trending = trendingSnapshot,
                savedAtMs = trendingSavedAt,
                stickersTrending = stickersSnapshot,
                stickersSavedAtMs = stickersSavedAt,
                stickersKind = stickersTab,
            ),
        )
    }

    fun fetch(pageQuery: String, append: Boolean) {
        if (append && (loading || loadingMore || !hasMore)) return
        val mine = ++generation
        if (append) loadingMore = true else loading = true
        error = ""
        scope.launch {
            try {
                val offset = if (append) nextOffset else 0
                val body = withContext(Dispatchers.IO) {
                    client.newCall(
                        Request.Builder()
                            .url(
                                GifPickerContract.buildUrl(
                                    GifPickerContract.DEFAULT_API_KEY, pageQuery.trim(), offset,
                                    stickers = stickersTab,
                                ),
                            )
                            .build(),
                    ).execute().use { response ->
                        if (!response.isSuccessful) throw IllegalStateException("Giphy ${response.code}")
                        response.body?.string() ?: throw IllegalStateException("Giphy empty body")
                    }
                }
                val fetched = GifPickerContract.parseChoices(body)
                val page = GifPickerContract.pagination(body, fetched.size, offset)
                if (mine != generation) return@launch
                items = if (append) items + fetched.filter { fresh -> items.none { it.id == fresh.id } } else fetched
                nextOffset = page.nextOffset
                hasMore = page.hasMore
                if (pageQuery.isBlank()) {
                    if (stickersTab) {
                        stickersSnapshot = items
                        stickersSavedAt = System.currentTimeMillis()
                    } else {
                        trendingSnapshot = items
                        trendingSavedAt = System.currentTimeMillis()
                    }
                    persist()
                }
                if (items.isEmpty() && pageQuery.isNotBlank()) {
                    error = "No GIFs matched \"${pageQuery.trim()}\"."
                }
            } catch (failure: Exception) {
                if (mine == generation) error = "Couldn't load GIFs. Check your connection."
            } finally {
                if (mine == generation) {
                    loading = false
                    loadingMore = false
                }
            }
        }
    }

    fun pick(gif: GifChoice) {
        recent = GifPickerContract.mergeRecent(recent, gif)
        persist()
        onDismiss()
        onPick(gif)
    }

    // Restore once (24 h fresh cache seeds trending + recents), then the
    // first network page when the cache came up empty.
    LaunchedEffect(Unit) {
        val cache = store.load()
        if (cache != null) {
            recent = cache.recent
            // The remembered tab wins unless the caller asked for stickers.
            stickersTab = defaultStickers || cache.stickersKind
            stickersSnapshot = cache.stickersTrending
            stickersSavedAt = cache.stickersSavedAtMs
            val gifsFresh = GifPickerContract.isCacheFresh(cache.savedAtMs, System.currentTimeMillis())
            if (gifsFresh && cache.trending.isNotEmpty()) {
                trendingSnapshot = cache.trending
                trendingSavedAt = cache.savedAtMs
            }
            val stickersFresh = GifPickerContract.isCacheFresh(cache.stickersSavedAtMs, System.currentTimeMillis())
            val cachedStickers = if (stickersFresh) cache.stickersTrending else emptyList()
            val seed = if (stickersTab) cachedStickers else if (gifsFresh) cache.trending else emptyList()
            if (seed.isNotEmpty() && items.isEmpty()) {
                items = seed
                nextOffset = seed.size
            }
        }
        restored = true
        if (items.isEmpty()) fetch("", append = false)
    }
    // Kind switch: reset to that kind's page (cached seed or refetch).
    LaunchedEffect(stickersTab) {
        if (!restored) return@LaunchedEffect
        recentTab = false
        hasMore = true
        nextOffset = 0
        items = emptyList()
        val snapshot = if (stickersTab) stickersSnapshot else trendingSnapshot
        if (snapshot.isNotEmpty()) {
            items = snapshot
            nextOffset = snapshot.size
        } else {
            fetch("", append = false)
        }
    }
    // Search-as-you-type (legacy 350 ms debounce).
    LaunchedEffect(query) {
        if (!restored) return@LaunchedEffect
        delay(GifPickerContract.SEARCH_DEBOUNCE_MS)
        recentTab = false
        fetch(query, append = false)
    }

    val visibleItems = if (recentTab) recent else items

    ModalBottomSheet(onDismissRequest = onDismiss, containerColor = BitOSColors.surface) {
        Column(
            Modifier
                .fillMaxWidth()
                .fillMaxHeight(0.75f),
        ) {
            // Header: ONE row — small search field + the inline kind
            // segment (GIFs vs Stickers). Saves a full chrome row for the
            // grid; stickers = transparent cut-outs (web parity).
            Row(
                Modifier.padding(horizontal = BitOSSpacing.lg),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(BitOSSpacing.sm),
            ) {
                Box(Modifier.weight(1f)) {
                    // Small search field: compact 48 dp row + label-size
                    // text; the placeholder follows the active kind.
                    space.bitos.app.ui.components.BitosTextField(
                        value = query,
                        onValueChange = { query = it },
                        placeholder = if (stickersTab) "Search stickers…" else "Search GIFs",
                        singleLine = true,
                        compact = true,
                        leadingIcon = {
                            Icon(
                                AppIcons.Search,
                                contentDescription = null,
                                tint = BitOSColors.textSecondary,
                                modifier = Modifier.size(16.dp),
                            )
                        },
                        textStyle = MaterialTheme.typography.labelLarge.copy(color = BitOSColors.textPrimary),
                        modifier = Modifier.fillMaxWidth(),
                    )
                }
                Surface(
                    shape = RoundedCornerShape(BitOSSpacing.sm),
                    color = BitOSColors.surfaceElevated.copy(alpha = 0.35f),
                    // Same height as the compact search field (48 dp) —
                    // the header reads as ONE control row.
                    modifier = Modifier.height(48.dp),
                ) {
                    Row(
                        Modifier
                            .fillMaxHeight()
                            .padding(3.dp),
                        horizontalArrangement = Arrangement.spacedBy(2.dp),
                    ) {
                        KindPill("GIFs", selected = !stickersTab) { stickersTab = false }
                        KindPill("Stickers", selected = stickersTab) { stickersTab = true }
                    }
                }
                if (loading) {
                    CircularProgressIndicator(
                        strokeWidth = 2.dp,
                        color = BitOSColors.primary,
                        modifier = Modifier.size(16.dp),
                    )
                }
            }
            // Recent / Trending pills (only with recents — legacy parity).
            if (query.isBlank() && recent.isNotEmpty()) {
                Surface(
                    shape = RoundedCornerShape(BitOSSpacing.md),
                    color = BitOSColors.surfaceElevated.copy(alpha = 0.35f),
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = BitOSSpacing.lg, vertical = BitOSSpacing.sm)
                        .padding(all = BitOSSpacing.xs),
                ) {
                    Row {
                        TabPill("Recent", selected = recentTab, modifier = Modifier.weight(1f)) { recentTab = true }
                        Spacer(Modifier.width(BitOSSpacing.xs))
                        TabPill("Trending", selected = !recentTab, modifier = Modifier.weight(1f)) { recentTab = false }
                    }
                }
            }
            // Grid.
            Box(Modifier.weight(1f)) {
                when {
                    error.isNotEmpty() -> EmptyState(text = error)
                    visibleItems.isEmpty() && !loading -> EmptyState(text = if (recentTab) "No recent GIFs yet." else "No GIFs yet.")
                    else -> LazyVerticalGrid(
                        columns = GridCells.Fixed(3),
                        horizontalArrangement = Arrangement.spacedBy(BitOSSpacing.sm),
                        verticalArrangement = Arrangement.spacedBy(BitOSSpacing.sm),
                        contentPadding = androidx.compose.foundation.layout.PaddingValues(
                            start = BitOSSpacing.lg,
                            end = BitOSSpacing.lg,
                            top = BitOSSpacing.sm,
                            bottom = BitOSSpacing.md,
                        ),
                        modifier = Modifier.fillMaxSize(),
                    ) {
                        items(visibleItems, key = { if (it.id.isEmpty()) it.url else it.id }) { gif ->
                            Box(
                                Modifier
                                    .aspectRatio(1f)
                                    .clip(RoundedCornerShape(BitOSSpacing.sm))
                                    .background(BitOSColors.surfaceElevated)
                                    .clickable(onClickLabel = "Pick GIF") { pick(gif) },
                            ) {
                                GifPreviewTile(preview = gif.preview)
                            }
                        }
                        if (!recentTab && hasMore && visibleItems.isNotEmpty()) {
                            item(span = { GridItemSpan(3) }) {
                                Box(Modifier.fillMaxWidth().padding(vertical = BitOSSpacing.sm), contentAlignment = Alignment.Center) {
                                    OutlinedButton(
                                        onClick = { fetch(query, append = true) },
                                        enabled = !loadingMore,
                                    ) {
                                        if (loadingMore) {
                                            CircularProgressIndicator(strokeWidth = 2.dp, modifier = Modifier.size(14.dp))
                                        } else {
                                            Text("Load more")
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
            // Powered by Giphy footer (web parity).
            Row(
                Modifier.fillMaxWidth().padding(vertical = 6.dp),
                horizontalArrangement = Arrangement.Center,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Icon(AppIcons.Chat, contentDescription = null, tint = BitOSColors.textTertiary, modifier = Modifier.size(12.dp))
                Spacer(Modifier.width(4.dp))
                Text("Powered by Giphy", style = MaterialTheme.typography.labelSmall, fontSize = 10.sp, color = BitOSColors.textTertiary)
            }
        }
    }
}

@Composable
private fun TabPill(label: String, selected: Boolean, modifier: Modifier = Modifier, onTap: () -> Unit) {
    Box(
        modifier
            .clip(RoundedCornerShape(BitOSSpacing.sm))
            // The app's chip language: accent-tint fill + accent text when
            // selected (surface-on-surface was near-invisible).
            .background(if (selected) BitOSColors.primary.copy(alpha = 0.12f) else androidx.compose.ui.graphics.Color.Transparent)
            .clickable(onClickLabel = label) { onTap() }
            // Medium tab: comfortable touch height + labelMedium text.
            .padding(horizontal = BitOSSpacing.sm, vertical = BitOSSpacing.sm)
            .semantics { contentDescription = label },
        contentAlignment = Alignment.Center,
    ) {
        Text(
            label,
            style = MaterialTheme.typography.labelMedium,
            fontWeight = if (selected) FontWeight.W600 else FontWeight.W400,
            color = if (selected) BitOSColors.primary else BitOSColors.textSecondary,
        )
    }
}

/** Compact inline kind pill (GIFs/Stickers) trailing the search field —
 *  fills the 48 dp segment so the touch target is the whole pill. */
@Composable
private fun KindPill(label: String, selected: Boolean, onTap: () -> Unit) {
    Box(
        Modifier
            .fillMaxHeight()
            .clip(RoundedCornerShape(BitOSSpacing.xs))
            .background(if (selected) BitOSColors.primary.copy(alpha = 0.12f) else androidx.compose.ui.graphics.Color.Transparent)
            .clickable(onClickLabel = label) { onTap() }
            .padding(horizontal = BitOSSpacing.sm)
            .semantics { contentDescription = label },
        contentAlignment = Alignment.Center,
    ) {
        Text(
            label,
            style = MaterialTheme.typography.labelMedium,
            fontWeight = if (selected) FontWeight.W600 else FontWeight.W400,
            color = if (selected) BitOSColors.primary else BitOSColors.textSecondary,
        )
    }
}

@Composable
private fun GifPreviewTile(preview: String) {
    AsyncImage(
        model = preview,
        contentDescription = null,
        contentScale = ContentScale.Crop,
        modifier = Modifier.fillMaxSize(),
        placeholder = ColorPainter(BitOSColors.surfaceElevated),
        error = rememberVectorPainter(AppIcons.BrokenImage),
    )
}

@Composable
private fun EmptyState(text: String) {
    Column(
        Modifier.fillMaxSize().padding(BitOSSpacing.xl),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ) {
        Icon(AppIcons.Photo, contentDescription = null, tint = BitOSColors.textSecondary, modifier = Modifier.size(26.dp))
        Spacer(Modifier.height(BitOSSpacing.sm))
        Text(text, style = MaterialTheme.typography.bodySmall, color = BitOSColors.textSecondary)
    }
}

File: apps/ios/BitOS/Features/Create/GifPickerSheet.swift
import BusinessCore
import SwiftUI

/// One GIF from the picker — the full-resolution URL is embedded in the
/// note, the preview renders the grid tile (shared `GifPickerContract`
/// item JSON shape via the bridge).
struct GifChoiceItem: Identifiable, Equatable {
    let id: String
    let url: String
    let preview: String
    let width: Int
    let height: Int

    var idKey: String { id.isEmpty ? url : id }

    static func from(jsonArray: [[String: Any]]) -> [GifChoiceItem] {
        jsonArray.compactMap { obj in
            guard let url = obj["url"] as? String,
                  let preview = obj["preview"] as? String else { return nil }
            return GifChoiceItem(
                id: (obj["id"] as? String) ?? "",
                url: url,
                preview: preview,
                width: (obj["w"] as? Int) ?? 120,
                height: (obj["h"] as? Int) ?? 120
            )
        }
    }

    var itemJson: String {
        let obj: [String: Any] = ["id": id, "url": url, "preview": preview, "w": width, "h": height]
        guard let data = try? JSONSerialization.data(withJSONObject: obj),
              let json = String(data: data, encoding: .utf8) else { return "{}" }
        return json
    }
}

/**
 * APP-008 GIF picker (legacy Flutter `GifPickerSheet` / web
 * `GifPicker.svelte` parity): trending on open, 350 ms debounced search,
 * Recent tab (≤12, persisted), 24 h trending cache, "Load more"
 * pagination, two-column preview grid and the "Powered by Giphy" footer.
 * All rules run in shared `GifPickerContract` through the bridge; this
 * view owns HTTP, tiles and persistence only.
 */
struct GifPickerSheet: View {
    let onPick: (GifChoiceItem) -> Void
    let onDismiss: () -> Void
    /// Open on the STICKERS tab (transparent cut-outs — what meme layers
    /// want on top of the media; the composer default stays GIFs).
    var defaultStickers: Bool = false

    private let bridge = BusinessCoreBridge()
    private let cacheKey = "bitos_gif_picker_v1"

    @State private var query = ""
    @State private var items: [GifChoiceItem] = []
    @State private var trendingSnapshot: [GifChoiceItem] = []
    @State private var trendingSavedAt: Double = 0
    // Web parity: GIFs vs Stickers — Giphy's sticker endpoints return
    // transparent cut-outs. Each kind keeps its own trending snapshot.
    @State private var stickersTab = false
    @State private var stickersSnapshot: [GifChoiceItem] = []
    @State private var stickersSavedAt: Double = 0
    @State private var recent: [GifChoiceItem] = []
    @State private var recentTab = false
    @State private var loading = false
    @State private var loadingMore = false
    @State private var errorText = ""
    @State private var nextOffset = 0
    @State private var hasMore = true
    @State private var restored = false
    @State private var generation = 0

    private var visibleItems: [GifChoiceItem] { recentTab ? recent : items }

    var body: some View {
        VStack(spacing: 0) {
            searchHeader
            if query.trimmingCharacters(in: .whitespaces).isEmpty && !recent.isEmpty {
                tabPills
            }
            grid
            footer
        }
        .background(BitOSTheme.surface)
        .task { restoreAndLoad() }
        // Search-as-you-type: .task(id:) cancels the previous debounce
        // when the query changes (shared 350 ms rule).
        .task(id: query) {
            guard restored else { return }
            try? await Task.sleep(nanoseconds: UInt64(bridge.gifPickerDebounceMs()) * 1_000_000)
            guard !Task.isCancelled else { return }
            recentTab = false
            await fetch(query, append: false)
        }
        .onChange(of: stickersTab) { _, _ in
            guard restored else { return }
            recentTab = false
            hasMore = true
            nextOffset = 0
            items = []
            let snapshot = stickersTab ? stickersSnapshot : trendingSnapshot
            if !snapshot.isEmpty {
                items = snapshot
                nextOffset = snapshot.count
            } else {
                Task { await fetch("", append: false) }
            }
        }
    }

    // MARK: - Pieces

    /// ONE header row: the small search field + the inline GIFs/Stickers
    /// segment (saves a full chrome row for the grid); the placeholder
    /// follows the active kind.
    private var searchHeader: some View {
        HStack(spacing: BitOSTheme.Spacing.sm) {
            HStack(spacing: BitOSTheme.Spacing.sm) {
                AppIcons.image(for: AppIcons.search)
                    .font(.system(size: 12))
                    .foregroundStyle(BitOSTheme.textSecondary)
                TextField(
                    stickersTab ? "Search stickers…" : "Search GIFs",
                    text: $query
                )
                .textInputAutocapitalization(.never)
                .autocorrectionDisabled()
                .font(.system(size: 13))
                .foregroundStyle(BitOSTheme.textPrimary)
                if loading {
                    ProgressView()
                        .controlSize(.small)
                        .tint(BitOSTheme.accent)
                }
            }
            .padding(.horizontal, 10)
            .frame(maxWidth: .infinity)
            .frame(height: 34)
            .background(
                RoundedRectangle(cornerRadius: BitOSTheme.Radius.md, style: .continuous)
                    .fill(BitOSTheme.surfaceElevated.opacity(0.45))
            )
            kindSegment
        }
        .padding(.horizontal, BitOSTheme.Spacing.lg)
        .padding(.top, BitOSTheme.Spacing.md)
        .padding(.bottom, BitOSTheme.Spacing.sm)
    }

    /// Compact trailing GIFs/Stickers segment — the SAME height as the
    /// search field (34 pt): the header reads as ONE control row.
    private var kindSegment: some View {
        HStack(spacing: 2) {
            kindPill("GIFs", selected: !stickersTab) { stickersTab = false }
            kindPill("Stickers", selected: stickersTab) { stickersTab = true }
        }
        .padding(3)
        .frame(height: 34)
        .background(
            RoundedRectangle(cornerRadius: BitOSTheme.Radius.sm, style: .continuous)
                .fill(BitOSTheme.surfaceElevated.opacity(0.35))
        )
    }

    /// Kind pill fills the segment height — the touch target is the pill.
    private func kindPill(_ label: String, selected: Bool, onTap: @escaping () -> Void) -> some View {
        Button(action: onTap) {
            Text(label)
                .font(.system(size: 13, weight: selected ? .semibold : .regular))
                .foregroundStyle(selected ? BitOSTheme.accent : BitOSTheme.textSecondary)
                .frame(maxHeight: .infinity)
                .padding(.horizontal, BitOSTheme.Spacing.sm)
                .background(
                    RoundedRectangle(cornerRadius: BitOSTheme.Radius.xs, style: .continuous)
                        .fill(selected ? BitOSTheme.accent.opacity(0.18) : .clear)
                )
        }
        .buttonStyle(.plain)
        .accessibilityLabel(label)
        .accessibilityAddTraits(selected ? [.isSelected] : [])
    }

    private var tabPills: some View {
        HStack(spacing: BitOSTheme.Spacing.xs) {
            tabPill("Recent", selected: recentTab) { recentTab = true }
            tabPill("Trending", selected: !recentTab) { recentTab = false }
        }
        .padding(BitOSTheme.Spacing.xs)
        .background(
            RoundedRectangle(cornerRadius: BitOSTheme.Radius.md, style: .continuous)
                .fill(BitOSTheme.surfaceElevated.opacity(0.35))
        )
        .padding(.horizontal, BitOSTheme.Spacing.lg)
        .padding(.bottom, BitOSTheme.Spacing.sm)
    }

    /// Medium tab: 13 pt semibold, comfortable touch height, and the
    /// app's accent chip language for the selected state (surface-on-
    /// surface was near-invisible).
    private func tabPill(_ label: String, selected: Bool, onTap: @escaping () -> Void) -> some View {
        Button(action: onTap) {
            Text(label)
                .font(.system(size: 13, weight: selected ? .semibold : .regular))
                .foregroundStyle(selected ? BitOSTheme.accent : BitOSTheme.textSecondary)
                .frame(maxWidth: .infinity)
                .padding(.vertical, BitOSTheme.Spacing.sm)
                .background(
                    RoundedRectangle(cornerRadius: BitOSTheme.Radius.sm, style: .continuous)
                        .fill(selected ? BitOSTheme.accent.opacity(0.18) : .clear)
                )
        }
        .buttonStyle(.plain)
        .accessibilityLabel(label)
        .accessibilityAddTraits(selected ? [.isSelected] : [])
    }

    private var grid: some View {
        Group {
            if !errorText.isEmpty {
                emptyState(icon: AppIcons.photo, text: errorText)
            } else if visibleItems.isEmpty && !loading {
                emptyState(icon: AppIcons.photo, text: recentTab ? "No recent GIFs yet." : "No GIFs yet.")
            } else {
                ScrollView {
                    LazyVGrid(columns: [GridItem(.flexible(), spacing: BitOSTheme.Spacing.sm), GridItem(.flexible(), spacing: BitOSTheme.Spacing.sm)], spacing: BitOSTheme.Spacing.sm) {
                        ForEach(visibleItems, id: \.idKey) { gif in
                            Button {
                                pick(gif)
                            } label: {
                                GifPreviewTile(preview: gif.preview)
                                    .aspectRatio(1, contentMode: .fit)
                                    .clipShape(RoundedRectangle(cornerRadius: BitOSTheme.Radius.sm, style: .continuous))
                            }
                            .buttonStyle(.plain)
                            .accessibilityLabel("Pick GIF")
                        }
                        if !recentTab && hasMore && !visibleItems.isEmpty {
                            Button {
                                Task { await fetch(query, append: true) }
                            } label: {
                                Group {
                                    if loadingMore {
                                        ProgressView().controlSize(.small)
                                    } else {
                                        Text("Load more")
                                    }
                                }
                                .frame(maxWidth: .infinity)
                                .padding(.vertical, BitOSTheme.Spacing.sm)
                            }
                            .buttonStyle(.bordered)
                            .disabled(loadingMore)
                            .gridCellColumns(2)
                        }
                    }
                    .padding(.horizontal, BitOSTheme.Spacing.lg)
                    .padding(.top, BitOSTheme.Spacing.sm)
                    .padding(.bottom, BitOSTheme.Spacing.md)
                }
            }
        }
    }

    private var footer: some View {
        HStack(spacing: 4) {
            AppIcons.image(for: AppIcons.info)
                .font(.system(size: 10))
                .foregroundStyle(BitOSTheme.textTertiary)
            Text("Powered by Giphy")
                .font(.system(size: 10))
                .foregroundStyle(BitOSTheme.textTertiary)
        }
        .padding(.vertical, 6)
    }

    private func emptyState(icon symbol: String, text: String) -> some View {
        VStack(spacing: BitOSTheme.Spacing.sm) {
            AppIcons.image(for: symbol)
                .font(.system(size: 26))
                .foregroundStyle(BitOSTheme.textSecondary)
            Text(text)
                .font(.system(size: 12))
                .foregroundStyle(BitOSTheme.textSecondary)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }

    // MARK: - Data

    /// Bridge map values arrive as native Swift types ([String: Any]).
    private func intFrom(_ map: [String: Any], _ key: String) -> Int {
        (map[key] as? Int) ?? ((map[key] as? NSNumber)?.intValue ?? 0)
    }

    private func restoreAndLoad() {
        if let wire = UserDefaults.standard.string(forKey: cacheKey),
           let envelope = bridge.gifCacheDecode(json: wire),
           let data = envelope.data(using: .utf8),
           let cache = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            let nowMs = Int64(Date().timeIntervalSince1970 * 1000)
            recent = itemsFrom(cache["recent"])
            let gifsSavedAt = (cache["savedAt"] as? NSNumber)?.doubleValue ?? 0
            if bridge.gifCacheFresh(savedAtMs: Int64(gifsSavedAt), nowMs: nowMs) {
                let trending = itemsFrom(cache["trending"])
                trendingSnapshot = trending
                trendingSavedAt = gifsSavedAt
            }
            let stickersSavedAtValue = (cache["stickersSavedAt"] as? NSNumber)?.doubleValue ?? 0
            if bridge.gifCacheFresh(savedAtMs: Int64(stickersSavedAtValue), nowMs: nowMs) {
                stickersSnapshot = itemsFrom(cache["stickersTrending"])
                stickersSavedAt = stickersSavedAtValue
            }
            // The remembered tab wins unless the caller asked for stickers.
            stickersTab = defaultStickers || ((cache["stickersKind"] as? NSNumber)?.boolValue ?? false)
            let seed = stickersTab ? stickersSnapshot : trendingSnapshot
            if !seed.isEmpty && items.isEmpty {
                items = seed
                nextOffset = seed.count
            }
        }
        restored = true
        if items.isEmpty {
            Task { await fetch("", append: false) }
        }
    }

    private func itemsFrom(_ node: Any?) -> [GifChoiceItem] {
        guard let array = node as? [[String: Any]] else { return [] }
        return GifChoiceItem.from(jsonArray: array)
    }

    private func persist() {
        let recentJson = "[" + recent.map(\.itemJson).joined(separator: ",") + "]"
        let trendingJson = "[" + trendingSnapshot.map(\.itemJson).joined(separator: ",") + "]"
        let stickersJson = "[" + stickersSnapshot.map(\.itemJson).joined(separator: ",") + "]"
        let wire = bridge.gifCacheEncode(
            recentJson: recentJson,
            trendingJson: trendingJson,
            savedAtMs: Int64(trendingSavedAt),
            stickersTrendingJson: stickersJson,
            stickersSavedAtMs: Int64(stickersSavedAt),
            stickersKind: stickersTab
        )
        UserDefaults.standard.set(wire, forKey: cacheKey)
    }

    private func fetch(_ pageQuery: String, append: Bool) async {
        if append && (loading || loadingMore || !hasMore) { return }
        let mine = generation + 1
        generation = mine
        if append { loadingMore = true } else { loading = true }
        errorText = ""
        defer {
            if mine == generation {
                loading = false
                loadingMore = false
            }
        }
        do {
            let offset = append ? nextOffset : 0
            guard let url = URL(string: bridge.gifPickerUrl(query: pageQuery, offset: Int32(offset), stickers: stickersTab)) else {
                throw URLError(.badURL)
            }
            var request = URLRequest(url: url)
            request.timeoutInterval = 10
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
                throw URLError(.badServerResponse)
            }
            let body = String(data: data, encoding: .utf8) ?? ""
            let fetched = GifChoiceItem.from(jsonArray: parseItems(body))
            let page = bridge.gifPickerPagination(responseJson: body, fetchedCount: Int32(fetched.count), requestedOffset: Int32(offset))
            guard mine == generation else { return }
            if append {
                let known = Set(items.map(\.idKey))
                items += fetched.filter { !known.contains($0.idKey) }
            } else {
                items = fetched
            }
            let pageMap = page
            nextOffset = intFrom(pageMap, "nextOffset")
            hasMore = pageMap["hasMore"] as? Bool ?? false
            if pageQuery.trimmingCharacters(in: .whitespaces).isEmpty {
                if stickersTab {
                    stickersSnapshot = items
                    stickersSavedAt = Date().timeIntervalSince1970 * 1000
                } else {
                    trendingSnapshot = items
                    trendingSavedAt = Date().timeIntervalSince1970 * 1000
                }
                persist()
            }
            if items.isEmpty && !pageQuery.trimmingCharacters(in: .whitespaces).isEmpty {
                errorText = "No GIFs matched \"\(pageQuery.trimmingCharacters(in: .whitespaces))\"."
            }
        } catch {
            if mine == generation {
                errorText = "Couldn't load GIFs. Check your connection."
            }
        }
    }

    private func parseItems(_ body: String) -> [[String: Any]] {
        guard let itemsJson = bridge.gifPickerParse(responseJson: body) as String?,
              let data = itemsJson.data(using: .utf8),
              let array = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else { return [] }
        return array
    }

    private func pick(_ gif: GifChoiceItem) {
        let recentJson = "[" + recent.map(\.itemJson).joined(separator: ",") + "]"
        let merged = bridge.gifPickerMergeRecent(recentJson: recentJson, pickJson: gif.itemJson)
        recent = itemsFromJsonString(merged)
        persist()
        onDismiss()
        onPick(gif)
    }

    private func itemsFromJsonString(_ json: String) -> [GifChoiceItem] {
        guard let data = json.data(using: .utf8),
              let array = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else { return [] }
        return GifChoiceItem.from(jsonArray: array)
    }
}

/** Preview tile: first frame while loading, broken-image fallback (legacy parity). */
private struct GifPreviewTile: View {
    let preview: String

    var body: some View {
        AsyncImage(url: URL(string: preview)) { phase in
            switch phase {
            case .success(let image):
                image.resizable().scaledToFill()
            case .failure:
                ZStack {
                    BitOSTheme.surfaceElevated
                    AppIcons.image(for: AppIcons.brokenImage)
                        .font(.system(size: 20))
                        .foregroundStyle(BitOSTheme.textSecondary)
                }
            default:
                BitOSTheme.surfaceElevated
            }
        }
    }
}
